framework/src/utils/MooseUtils.C

```diff
@@ -349,6 +350,17 @@ baseName(const std::string & name)
   return name.substr(0, name.find_last_of('/') != std::string::npos ? name.find_last_of('/') : 0);
 }
 
+std::string
+hostname()
+{
+  // This is from: https://stackoverflow.com/a/505546
+  char hostname[1024];
+  hostname[1023] = '\0';
+  gethostname(hostname, 1023);
```

Comment:
Should probably always check return values of system calls...

Reply:
Good idea

Reply:
I just followed your stackoverflow link. It says that gethostname is deprecated and to use getaddrinfo. Maybe you could see if that works instead. Could probably also switch this to std::array instead of raw c-arrays if you are going to modernize it.

Reply:
NVM - you aren't even using that function.

Reply:
Hah - I was just getting ready to say that :-)  It's still very much alive according to the Open Group: http://pubs.opengroup.org/onlinepubs/9699919799/

I'll give `std::array` a shot - good practice anyway :-)

Reply:
Why didn't they provide casting operators for `std::array`?

```
 no known conversion from 'std::array<char, 1024>' to 'char *'
```

dumb!

Reply:
Also this:

```
no viable conversion from returned value of type 'std::array<char, 1024>' to function return type 'std::string'
```

So here is the dumb new "modern" version of this function:

```
std::string
hostname()
{
  // Mostly from: https://stackoverflow.com/a/505546
  std::array<char, 1024> hostname;
  hostname[1023] = '\0';
  auto failure = gethostname(hostname.data(), 1023);

  if (failure)
    mooseError("Failed to retrieve hostname!");

  return std::string(hostname.data());
}
```

I like the original better - I'm going back (keeping the error checking of course).

Reply:
Well this isn't the best you could do with a modern version. `std::array` is typesafe with several methods for querying and manipulating the data. In your version, you still have way too many magic numbers. You can get the last item with `.back()` to set the null byte. You can also get the `.size()`. Finally, as you figured out, you can get a pointer to the underlying raw data with `.data()`. It makes sense that you can't cast it, the type has extra data associated with it so a cast shouldn't work, instead you just ask for access to the underlying data buffer (like you do with `std::vector`). I think if you used all of the newer methods, it would look better, but whatever. Like a raw c-array, it's stack allocated so it is superior to std::vector when the size is fixed.

Reply:
I still think they should have made it auto cast to a `T*`... that would make it WAY easier to interoperate with C functions.

In addition, I think a constructor for `std::string` that takes a `std::array<char>` makes complete sense... or, again, it would automatically work if the `std::array<char>` would auto-cast to a `char *`...

But whatever: we all know that I like APIs for usability... I'm sure @rwcarlsen would _love_ this type of API purity.  I can just hear him now: "But having all of the casts be explicit is way better because..." (and I'm sure he could list 6 reasons why it's a good idea) 😃 

In this case: the whole purpose of this function is simply to call a C function... I think it makes sense to do it in the C way... it certainly makes the code nice, simple and clean.